### PR TITLE
HPCC-18308 eclcmd should pass eclcc options as debug values

### PR DESCRIPTION
**Pull request by afishbeck**
"ecl run <target> <file> -f-<option>"

Allows arbitrary eclcc options to be passed to the eclcc command line.
Those options should also be passed as workunit debug values formatted
as "eclcc-<option>".

Signed-off-by: Anthony Fishbeck <anthony.fishbeck@lexisnexis.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

### CODE REVIEW

**Comment by richardkchapman** [comment]
@ghalliday Please sanity check this

File: ecl/eclcmd/eclcmd_core.cpp [hunk 1]
```diff
@@ -148,6 +148,8 @@ bool doDeploy(EclCmdWithEclTarget &cmd, IClientWsWorkunits *client, const char *
     expandDefintionsAsDebugValues(cmd.definitions, cmd.debugValues);
     if (cmd.debugValues.length())
     {
+        IArrayOf<IEspNamedValue> debugValues;
```
Comment (ghalliday) [inline ecl/eclcmd/eclcmd_core.cpp]: trivial: not used.

File: ecl/eclccserver/eclccserver.cpp [hunk 2]
```diff
@@ -254,7 +254,8 @@ class EclccCompileThread : implements IPooledThread, implements IErrorReporter,
         {
             //Allow eclcc-xx-<n> so that multiple values can be passed through for the same named debug symbol
             const char * start = option + (*option=='-' ? 1 : 6);
-            const char * dash = strrchr(start, '-');     // position of second dash, if present
+            const char * finger = (*start=='-') ? start+1 : start; //support leading double dash
```
Comment (ghalliday) [inline ecl/eclccserver/eclccserver.cpp]: I don't understand this change - can you elaborate?

**Comment by ghalliday** [comment]
@afishbeck I may be misunderstanding, but it looks like an option -f-x is converted to eclcc-x, but would then be matched in eclccserver as optionx.  I would have expected it to be converted to eclcc--x (then the double minus code would make sense.)

File: ecl/eclccserver/eclccserver.cpp (hunk 2 above)
Comment (afishbeck) [reply ecl/eclccserver/eclccserver.cpp]: @ghalliday It's slightly confusing, but the key is that when the option is appended to the eclcc command line the stripped '-' is restored.  I was trying not to refactor the code more than I had to as this is a last minute PR.

`eclccCmd.appendf(" -%s=%s", optName.get(), value);`

Some eclcc options actually have two dashes.  "--nostdinc" for example.  So this change preserves the second dash in the option name "-nostdinc" and then both dashes end up on the command line.

**Comment by afishbeck** [comment]
@ghalliday Removed unused variable.  Responded to your other comment.

**Comment by ghalliday** [comment]
@richardkchapman looks good.  I'll let you decide which build it goes in.

**Comment by richardkchapman** [comment]
I have a concern that --fastparse and --nostdinc should be being applied to the creation of the archive when ecl command calls out to a local eclcc, but don't necessarily apply to the parsing at the remote end - so while it's ok to fix the -f- processing as you have here, we really need --fastparse and --nostdinc to be also handled explicitly by all ecl commands.

I also suspect we may need to push this out to 6.4.4. There is a workaround of creating the archive independently I think.

**Comment by afishbeck** [comment]
@richardkchapman I could easily add those as explicit eclcmd options.

My guess would have been that --nostdinc had no effect on the backend since the depencies were already gathered into the archive at the client side?

I can't seem to find --fastparse.

**Comment by afishbeck** [comment]
As mentioned in the meeting, --fastparse was a typo, it's --fastsyntax.

**Pull request merged**